code/modules/holomap/holomap.dm

```diff
@@ -0,0 +1,500 @@
+	//use sparingly, dont make regions different only by a few values, just make it a single one across both instances (they are very costly to make and store)
+	/// should be a list {start x, start y, end x, end y} start should be bottom left corner; end should be top right... everything inside gets rendered (out of 255)
+	var/list/region_selection = list(1, 1, 255, 255)
+	/// this holds all the people it is currently displaying to
+	var/list/display_targets = list()
+	/// the screen obj that holds the image for your viewing pleasure
+	var/datum/holomap/my_map
+	///action for displaying a location to the user
+	var/list/actions_users = list()
+
+/obj/machinery/holomap/Initialize(mapload)
+	. = ..()
+
+	if(!zlevel_rendered)
+		zlevel_rendered = z
+
+	my_map = new /datum/holomap(region_selection, zlevel_rendered, x, y, z)
+
+/obj/machinery/holomap/proc/show_location(mob/user)
+	if(user.client && actions_users[user.key])
+		var/datum/action/show_location/SL = actions_users[user.key]
+		user.client.screen += SL.location_displayed
+
+/obj/machinery/holomap/attack_hand(mob/user)
+	. = ..()
+	if(stat & (NOPOWER|BROKEN) || user.incapacitated() || !user.Adjacent(src))
+		return
+
+	if(!ishuman(user))
+		atom_say("Unable to locate subject's optical organ. Place optical organ infront of device and try again.")
+		return
+
+	if(emagged)
+		var/mob/living/carbon/M = user
+		M.flash_eyes(2, 1)
+	if(user in display_targets)
+		stopdisplaymap(user)
+	else
+		displaymap(user)
+
+/obj/machinery/holomap/proc/displaymap(mob/M)
+	if(!M.client)
+		return
+	display_targets |= M
+	M.client.screen += my_map.using
+	M.client.screen += my_map.arrow
+	var/datum/action/show_location/SL = new /datum/action/show_location(src)
+	actions_users[M.key] = SL
+	SL.Grant(M)
+	update_icon(UPDATE_ICON)
+	if(do_after(M, 5 MINUTES, FALSE, src, FALSE))
+		atom_say("<span class='warning'>Maximum display time reached, shutting down session.</span>")
+		stopdisplaymap(M)
+	else
+		stopdisplaymap(M)
+
+/obj/machinery/holomap/proc/stopdisplaymap(mob/user)
+	display_targets -= user
+	user.client.screen -= my_map.using
+	user.client.screen -= my_map.arrow
+	var/datum/action/show_location/SL
+	if(actions_users[user.key])
+		SL = actions_users[user.key]
+		user.client.screen -= SL.location_displayed
+		SL.Remove(user)
+	actions_users -= user.key
+	if(SL)
+		qdel(SL)
+	update_icon(UPDATE_ICON)
+
+/obj/machinery/holomap/update_icon()
+	. = ..()
+
+	if(stat & (NOPOWER|BROKEN))
+		icon_state = "map_off"
+		return
+
+	if(length(display_targets))
+		icon_state = "map_projecting"
+	else if(emagged)
+		icon_state = "map_emag"
+	else
+		icon_state = "map"
+
+/obj/machinery/holomap/emag_act(user as mob)
+	emagged = TRUE
+	to_chat(user, "You short out the emission regulators of [src]!")
+	for(var/mob/living/carbon/M in display_targets)
+		M.flash_eyes(2, 1)
+
+/obj/machinery/holomap/power_change()
+	. = ..()
+	update_icon()
+
+/obj/screen/holomap/proc/make_map()
```

Comment:
ok bear with me me, we're gonna micro optimize the shit out of this proc. This bad boy is expensive as all fucking hell, even more than I previously thought.
Within this proc as it is. You will iterate over a little more than 65 thousand turfs which will look through 4 cardinal direction and 4 diagonal direction per cardinal direction. I've gone through the entire proc and identified the amount of calls we make the can just be completely cut out probably:

- 21 get_area() calls per turf (1.3 million calls over all) which can be simplified to `turf.loc`
- 23 unnecessary istype() calls per turf (1.5 million calls over all), some are in get_area() others are just repeated istypes that can be cached
- 21 get_turf() calls which are completely redundant and can be cut out by removing get_area()
- 21 get_step() calls, again, get_area()
- You could also overall, cut out like 100 istype checks by rearranging when you check windows. This is overall and not perf turf like the others

The grand total of unnecessary proc calls is.... *drumroll*.... 5.6 million :D. While an on-station map may only get to about half this mark, the lavaland holomap definitely will hit this number.

Reply:
My main goal was just to get the correct icons rendered, and not optimization.. since I knew this was probably only going to get run once. But all said I am happy to go through it all to make it more optimized.

Reply:
I did as well as I could and cleaned up all mentioned.